src/main/java/org/opensearch/neuralsearch/query/NeuralSparseQueryBuilder.java

```diff
@@ -270,40 +304,56 @@ private static void parseQueryParams(XContentParser parser, NeuralSparseQueryBui
     }
 
     @Override
-    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
-        // We need to inference the sentence to get the queryTokens. The logic is similar to NeuralQueryBuilder
-        // If the inference is finished, then rewrite to self and call doToQuery, otherwise, continue doRewrite
-        if (null != queryTokensSupplier) {
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        if (Objects.nonNull(queryTokensSupplier)) {
             return this;
         }
-
         validateForRewrite(queryText, modelId);
         SetOnce<Map<String, Float>> queryTokensSetOnce = new SetOnce<>();
-        queryRewriteContext.registerAsyncAction(
-            ((client, actionListener) -> ML_CLIENT.inferenceSentencesWithMapResult(
-                modelId(),
-                List.of(queryText),
-                ActionListener.wrap(mapResultList -> {
-                    queryTokensSetOnce.set(TokenWeightUtil.fetchListOfTokenWeightMap(mapResultList).get(0));
-                    actionListener.onResponse(null);
-                }, actionListener::onFailure)
-            ))
-        );
+        queryRewriteContext.registerAsyncAction(getModelInferenceAsync(queryTokensSetOnce));
         return new NeuralSparseQueryBuilder().fieldName(fieldName)
             .queryText(queryText)
+            .queryName(queryName)
+            .boost(boost)
             .modelId(modelId)
             .maxTokenScore(maxTokenScore)
-            .queryTokensSupplier(queryTokensSetOnce::get);
+            .queryTokensSupplier(queryTokensSetOnce::get)
+            .twoPhaseSharedQueryToken(twoPhaseSharedQueryToken)
+            .twoPhasePruneRatio(twoPhasePruneRatio);
+    }
+
+    private BiConsumer<Client, ActionListener<?>> getModelInferenceAsync(SetOnce<Map<String, Float>> setOnce) {
+        return ((client, actionListener) -> ML_CLIENT.inferenceSentencesWithMapResult(
+            modelId(),
+            List.of(queryText),
+            ActionListener.wrap(mapResultList -> {
+                Map<String, Float> queryTokens = TokenWeightUtil.fetchListOfTokenWeightMap(mapResultList).get(0);
+                if (Objects.nonNull(twoPhaseSharedQueryToken)) {
+                    Map<String, SetOnce<Map<String, Float>>> splitSetOnce = splitQueryTokensByRatioedMaxScoreAsThreshold(
+                        queryTokens,
+                        twoPhasePruneRatio
+                    );
+                    setOnce.set(splitSetOnce.get(UP_THRESHOLD).get());
+                    twoPhaseSharedQueryToken.putAll(splitSetOnce.get(DOWN_THRESHOLD).get());
```

Comment:
Can we just assign the result Map to twoPhaseSharedQueryToken? The putAll needs to put all elements with o(N) complexity

Reply:
Ok, I make it into assign now.

Reply:
Seems unneccessary to wrap it in a brace like `(splitQueryTokens.v2())`

Reply:
remove the redundant.